[Core] Add default Ray Node labels at Node init

ryanaoleary opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

This PR adds support for populating several default Ray node labels (described [here](https://docs.google.com/document/d/1DKhPuZERlLbsU4TIQZZ6POCsm1pVMBgN_yn5r0OmaDI/edit?tab=t.0#heading=h.zaw235b8zg9p)) in the Ray runtime environment when a node is initialized. This change will help support autoscaling with the Label Selector API. This PR is related to https://github.com/ray-project/kuberay/pull/3699 which passes several environment variables from the K8s stack which are used to set `ray.io/` labels. I'll leave a comment on this PR with manual tests showing the `ray.io/accelerator-type` and `ray.io/availability-zone` labels getting set.

## Related issue number

https://github.com/ray-project/ray/issues/51564

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

ryanaoleary committed:
Add default Ray Node labels at Node init

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

Add default Ray Node labels at Node init

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

Add default Ray Node labels at Node init

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Fix test

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary commented:
cc: @MengjinYan

MengjinYan commented on python/ray/tests/test_node_labels.py:
```diff
@@ -152,13 +154,42 @@ def test_ray_start_set_node_labels_from_file():
+    assert labels.get("ray.io/node-group") == "worker-group-1"
+    assert labels.get("ray.io/availability-region") == "us-central2"
+    assert labels.get("ray.io/availability-zone") == "us-central2-b"
+    assert (
```
Wondering if it is possible to test the `accelerator-type` by mocking the `ResourceSpec` or input the accelerator type through custom resources?

ryanaoleary replied:
I must've just been doing something wrong before - I tried again and was able to get it to detect a TPU accelerator just by setting the env var. We now actually test `accelerator-type` in [bd61ea6](https://github.com/ray-project/ray/pull/53360/commits/bd61ea662e21725ce78aa6f70596be660e997473).

MengjinYan reviewed (commented):
The logic change looks good to me! Just one comment about testing.

ryanaoleary committed:
Fix ray re-init error

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Actually check accelerator-type in test

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

jjyao commented on python/ray/_private/node.py:
```diff
@@ -1940,3 +1942,36 @@ def _record_stats(self):
+        node_labels = self._get_node_labels()
+
+        # Get environment variables populated from K8s Pod Spec
+        market_type = os.environ.get("RAY_NODE_MARKET_TYPE", "")
```
so they cannot be passed in as `ray start --labels`?

ryanaoleary replied:
For things like node-group we could pass it to `ray start --labels`, but other labels like `region`/`zone` and `accelerator-type` aren't known until the Pod is scheduled (and for the latter when the node is actually initialized), which occurs after the Pod spec is built by KubeRay. The Ray node group is currently already passed as an environment variable by KubeRay [here](https://github.com/ray-project/kuberay/blob/f460fda7ef6369d00feb6f2f5b919dd73074dc57/ray-operator/controllers/ray/common/pod.go#L586), so I figured we could just re-use it rather than lookup the Pod labels.

edoakes commented on python/ray/tests/test_node_labels.py:
```diff
@@ -152,13 +154,44 @@ def test_ray_start_set_node_labels_from_file():
+    assert labels.get("ray.io/availability-zone") == "us-central2-b"
+    assert labels.get("ray.io/accelerator-type") == "TPU-V4"
+
+    # Unset env vars
+    for k in [
+        "RAY_NODE_MARKET_TYPE",
+        "RAY_NODE_TYPE_NAME",
+        "RAY_NODE_ZONE",
+        "RAY_NODE_REGION",
+        "TPU_ACCELERATOR_TYPE",
+    ]:
+        del os.environ[k]
```
this needs to be done in a context manager or fixture, else if the test fails it will leak global process state

ryanaoleary replied:
Moved it to a test fixture that sets and then clears the env vars under test in [881932d](https://github.com/ray-project/ray/pull/53360/commits/881932dfb179ba8a48f5be8c941fdcfdda21b313).

ryanaoleary committed:
Move Ray node related keys/env vars to constants.h

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Add test fixture

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Fix comments

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary commented on python/ray/_private/node.py:
```diff
@@ -1940,3 +1942,36 @@ def _record_stats(self):
             # to avoid any issues.
             record_hardware_usage(cpu_model_name[:50])
+
+    def _add_default_ray_node_labels(self, resource_spec):
```
I changed it to just call `get_resource_spec` in the helper function in 
33d3fd2, so this isn't passed as an arg anymore.

ryanaoleary commented on python/ray/_private/node.py:
```diff
@@ -1940,3 +1942,36 @@ def _record_stats(self):
+
+        # Get accelerator type from ResourceSpec
+        accelerator_type = None
+        if resource_spec.resolved():
```
Left a comment in 33d3fd2, it's just checking that `num_cpus`, `num_gpus`, `memory`, `object_store_memory`, and `resources` have been set in the resource configuration. We need this check because to get the acceleratory type we rely on calling `accelerator_manager.get_current_node_accelerator_type()` and checking that for the given resource, `resource_spec.resources[<RESOURCE-NAME>]` is not None.

ryanaoleary commented on python/ray/_private/node.py:
```diff
@@ -1940,3 +1942,36 @@ def _record_stats(self):
+        if availability_zone:
+            node_labels[ray._raylet.RAY_NODE_ZONE_KEY] = availability_zone
+        if availability_region:
+            node_labels[ray._raylet.RAY_NODE_REGION_KEY] = availability_region
```
Done in 33d3fd2, we now merge the dictionary of labels passed in to the Ray params and dictionary of labels returned from `_get_default_ray_node_labels`, with the values passed in by the user taking precedence.

ryanaoleary commented on python/ray/_private/node.py:
```diff
@@ -1940,3 +1942,36 @@ def _record_stats(self):
             record_hardware_usage(cpu_model_name[:50])
+
+    def _add_default_ray_node_labels(self, resource_spec):
+        node_labels = self._get_node_labels()
```
Done in 33d3fd2, `_get_default_ray_node_labels` is now a helper function that returns a dictionary of default Ray labels, which we then merge with the labels passed in from the Ray params before passing the combined set of labels to the Raylet.

ryanaoleary committed:
Fix Ray data test

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

edoakes commented:
CI tests failures @ryanaoleary

ryanaoleary committed:
Fix exit_logging test and add try block

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary commented:
> CI tests failures @ryanaoleary

@edoakes Think I fixed the CI failure with [bd51036](https://github.com/ray-project/ray/pull/53360/commits/bd51036d2e3cffbad6969d7510f05a05c3fc4272). `test_exit_logging` was failing because the calls to `AcceleratorManager` in Node init would add the Traceback to the driver log, when none was expected. I updated the test to patch calls to `_get_default_ray_node_labels` so that these logs could be ignored for the test. An alternative would be to update calls like [this](https://github.com/ray-project/ray/blob/7337f2ad7315a587b5a86b86e244d7139b53bcc8/python/ray/_private/accelerators/npu.py#L80) in each `AcceleratorManager` definition to log but not add to the traceback. I also added a `try: except` block when we make calls to `AcceleratorManager` in the new code to ensure Node init doesn't fail for whatever reason when adding the default labels.

edoakes commented:
The solution looks worrisome/hacky. What causes the accelerator manager calls to fail on startup? We might be breaking some kind of assumption there.

@jjyao PTAL at the usage of accelerator manager here

ryanaoleary commented:
> The solution looks worrisome/hacky. What causes the accelerator manager calls to fail on startup? We might be breaking some kind of assumption there.
> 
> @jjyao PTAL at the usage of accelerator manager here

It'd raise an exception [here](https://github.com/ray-project/ray/blob/e35c2c80879b283cad64213bb829b6b7f00e411b/python/ray/_private/accelerators/npu.py#L81) due to missing dependencies such as `ModuleNotFoundError: No module named 'acl'` for NPU. Detecting the accelerator type would fail, but node startup would still succeed as expected. I added the `try: except` just to be safe since the code is on a critical path, but maybe it's not needed, since it's not used in other places `AcceleratorManager` is used. Maybe a better solution is just to `logger.debug` when we see a `ModuleNotFoundError` rather than raise an exception.

jjyao commented on src/ray/common/constants.h:
```diff
@@ -97,8 +101,13 @@ constexpr char kLibraryPathEnvName[] = "LD_LIBRARY_PATH";
 #define RAY_LABEL_KEY_PREFIX "ray.io/"
-/// Default node label key: node_id
+/// Default node label keys populated by the Raylet
 constexpr char kLabelKeyNodeID[] = RAY_LABEL_KEY_PREFIX "node_id";
```
For consistency: could you change `node_id` to `node-id` as well?

ryanaoleary replied:
Done in [b8ad0d3](https://github.com/ray-project/ray/pull/53360/commits/b8ad0d308a3b50ba571a8de903a28207efaf724b) and updated related test

jjyao commented on python/ray/_private/node.py:
```diff
@@ -1916,3 +1921,64 @@ def _record_stats(self):
+
+        # Get accelerator type from AcceleratorManager
+        if resource_spec.resolved():
+            # Check first that the resource configuration passed to the Raylet has been set.
+            # Only check for accelerator-type if resource_spec.resources is not None.
+            for (
+                accelerator_resource_name
+            ) in ray._private.accelerators.get_all_accelerator_resource_names():
+                try:
+                    accelerator_manager = (
+                        ray._private.accelerators.get_accelerator_manager_for_resource(
+                            accelerator_resource_name
+                        )
+                    )
+                    num_accelerators = resource_spec.resources.get(
+                        accelerator_resource_name, None
+                    )
+                    accelerator_type = (
+                        accelerator_manager.get_current_node_accelerator_type()
+                    )
+                    if num_accelerators is None:
+                        # Try to automatically detect the number of accelerators.
+                        num_accelerators = (
+                            accelerator_manager.get_current_node_num_accelerators()
+                        )
+                except Exception:
+                    accelerator_type = None
+                    logger.debug(
+                        "Failed to detect accelerator type for default Ray labels during Node init."
+                    )
+                if accelerator_type and num_accelerators:
+                    # Only set label if both accelerator type and the number of
+                    # accelerators were successfully detected.
+                    default_labels[
+                        ray._raylet.RAY_NODE_ACCELERATOR_TYPE_KEY
+                    ] = accelerator_type
+                    break  # Only add one value for ray.io/accelerator-type
```
This duplicates the logic of deciding accelerator type of the node inside `resource_spec.py`. After resource_spec is resolved, we should already know what the accelerator type is. We can just remember accelerator_type in ResourceSpec after it's resolved.

ryanaoleary replied:
The utility is a bit different in each location so I broke it into two helper functions in [57eabeb](https://github.com/ray-project/ray/pull/53360/commits/57eabebb6f4f33b59533938336dbbbbee8a360f6) with unit tests, we now call `resolve_and_update_accelerator_resources` in ResourceSpec and when the `accelerator_type` resource is removed, we can just remove this code block:

```
if accelerator_type:
   resources[f"{RESOURCE_CONSTRAINT_PREFIX}{accelerator_type}"] = 1

    from ray._private.usage import usage_lib

    usage_lib.record_hardware_usage(accelerator_type)
```
In node init we now call `get_first_detectable_accelerator_type` instead which just iterates through all the accelerator types and returns the first detected accelerator with `num_accelerators` set.

ryanaoleary committed:
Change node id constant

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Fix tests that use node_id

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Add accelerator util functions

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

edoakes commented on python/ray/_private/node.py:
```diff
@@ -1916,3 +1924,37 @@ def _record_stats(self):
+
+    # _get_default_ray_node_labels is a helper function to return a dictionary with
+    # default ray.io/ labels set for this node.
+    def _get_default_ray_node_labels(self):
```
This sounds like a good idea to me

ryanaoleary committed:
Fix comments and add ResourceAndLabelSpec

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Fix comment

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary commented on src/ray/common/constants.h:
```diff
@@ -97,8 +101,13 @@ constexpr char kLibraryPathEnvName[] = "LD_LIBRARY_PATH";
+/// Default node label keys populated by the Raylet
+constexpr char kLabelKeyNodeID[] = RAY_LABEL_KEY_PREFIX "node-id";
+constexpr char kLabelKeyNodeAcceleratorType[] = RAY_LABEL_KEY_PREFIX "accelerator-type";
+constexpr char kLabelKeyNodeMarketType[] = RAY_LABEL_KEY_PREFIX "market-type";
```
Done in https://github.com/ray-project/ray/commit/9ee715be783c11ed2e5f64590db53a238e2e82ae

ryanaoleary commented on python/ray/_private/node.py:
```diff
@@ -1916,3 +1924,37 @@ def _record_stats(self):
+    # default ray.io/ labels set for this node.
+    def _get_default_ray_node_labels(self):
+        default_labels = {}
+        resource_spec = self.get_resource_spec()
```
I had it there because we were using the resolved `resources` in the ResourceSpec to get the accelerator type, with the changes in  9ee715b we no longer do this since the logic is consolidated under `ResourceAndLabelSpec`

ryanaoleary commented on python/ray/_private/node.py:
```diff
@@ -1916,3 +1924,37 @@ def _record_stats(self):
+            default_labels[ray._raylet.RAY_NODE_REGION_KEY] = availability_region
+
+        # Get accelerator type from AcceleratorManager
+        if resource_spec.resolved():
```
Resolved in [9ee715b](https://github.com/ray-project/ray/pull/53360/commits/9ee715be783c11ed2e5f64590db53a238e2e82ae)

MengjinYan reviewed (commented):
All minor comments.

ryanaoleary committed:
Fix nits / rename file

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Fix Java test using node_id

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary commented on python/ray/_private/node.py:
```diff
@@ -290,7 +290,7 @@ def __init__(
             )
             # Set node labels from RayParams or environment override variables.
-            self._node_labels = self._get_node_labels()
+            # self._node_labels = self._get_node_labels()
```
Oh oops, done in [a215869](https://github.com/ray-project/ray/pull/53360/commits/a215869d8ba891dc70d1909f19e9b234b41a13f7)

ryanaoleary commented on python/ray/_private/services.py:
```diff
@@ -1658,8 +1656,9 @@ def start_raylet(
-    static_resources = resource_spec.to_resource_dict()
+    assert resource_and_label_spec.resolved()
+    static_resources = resource_and_label_spec.to_resource_dict()
+    labels = resource_and_label_spec.get_resolved_labels()
```
Yeah that works, in [a215869](https://github.com/ray-project/ray/pull/53360/commits/a215869d8ba891dc70d1909f19e9b234b41a13f7) I changed it to just directly get the labels from spec

ryanaoleary commented on python/ray/_private/resource_spec.py:
Sure, done in [a215869](https://github.com/ray-project/ray/pull/53360/commits/a215869d8ba891dc70d1909f19e9b234b41a13f7). I had to change the import path used in a few other files though which added some new code owners as reviewers

jjyao commented on python/ray/_private/accelerators/utils.py:
```diff
@@ -0,0 +1,103 @@
+from ray._private.accelerators import AcceleratorManager
+
+
+def get_current_node_accelerator(
```
def get_current_node_accelerator(num_gpus, resources) -> Optional[Tuple[AcceleratorManager, int]]

ryanaoleary replied:
Done in [e9c0ae1](https://github.com/ray-project/ray/pull/53360/commits/e9c0ae13b0c89d6f800d5879d7ec36e4aac576c6).

jjyao commented on python/ray/_private/resource_and_label_spec.py:
```diff
@@ -37,6 +43,8 @@ class ResourceSpec(
             Note that when calling to_resource_dict(), this will be scaled down
             by 30% to account for the global plasma LRU reserve.
         resources: The custom resources allocated for this raylet.
+        labels: The labels associated with this node. Labels can be used along
```
def resolve():
  accelerator_manager, num_accelerators = _get_current_node_accelerators()
   _resolve_resources(accelerator_manager, num_accelerators)
   _resolve_labels(accelerator_manager)

ryanaoleary replied:
Refactored to match this structure in [e9c0ae1](https://github.com/ray-project/ray/pull/53360/commits/e9c0ae13b0c89d6f800d5879d7ec36e4aac576c6).

ryanaoleary committed:
Refactor resolve functions

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

jjyao commented on python/ray/_private/accelerators/utils.py:
```diff
@@ -0,0 +1,37 @@
+        if resource_name == "GPU" and num_gpus:
+            num_accelerators = num_gpus
+
+        num_accelerators = resources.get(resource_name)
```
this would be in the else case, otherwise num_accelerators will be None for GPU

ryanaoleary replied:
Oh yeah nice catch, fixed in [ed1636f](https://github.com/ray-project/ray/pull/53360/commits/ed1636f03fc7b00fd244bd900625b544da370de8) to now correctly handle `non-gpus`.

jjyao commented on python/ray/_private/accelerators/utils.py:
```diff
@@ -0,0 +1,37 @@
+        if num_accelerators > 0:
+            return accelerator_manager, num_accelerators
+
+    return None, 0
```
Should just return None? Although I'm fine with changing the return type to be just `Tuple[AcceleratorManager, int]` and return (None, None) if there is no accelerator

ryanaoleary replied:
Done in [ed1636f](https://github.com/ray-project/ray/pull/53360/commits/ed1636f03fc7b00fd244bd900625b544da370de8)

jjyao commented on python/ray/_private/accelerators/utils.py:
```diff
@@ -0,0 +1,37 @@
+from ray._private.accelerators import AcceleratorManager
+
+
+def get_current_node_accelerator(
```
This is only used by the resource_and_label_spec, should we move it there and change it to be a private method?

ryanaoleary replied:
I think it makes sense to leave it in the `utils.py` file for now since it's specific to the `AcceleratorManager` class / accelerator interface and doesn't call anything specific to resolve resources or labels like in `ResourceAndLabelSpec`. I think this function may be used along with `TPUAcceleratorManager` for some of the planned TPU device discovery work. I don't have a strong preference though so I can also move it to `resource_and_label_spec` and try to keep any related logic in that class.

jjyao replied:
`accelerator interface` mainly cares about the physical accelerator resources but this function also takes into consideration of logical resources user explicitly set as well so that logic is specific to resolve resources.

jjyao commented on python/ray/_private/resource_and_label_spec.py:
```diff
@@ -272,12 +231,138 @@ def resolve(self, is_head: bool, node_ip_address: Optional[str] = None):
+
+        return merged
+
+    def resolve_and_update_accelerator_resources(
+        self, accelerator_manager, num_accelerators, resources: Dict[str, float]
+    ) -> int:
```
Its a bit odd that num_gpus is returned and other accelerator resources is added to argument resources.

Actually I think there is no good reason for `resolve()` to return a new ResourceSpec. We can just update self.

I feel this is better:

```
def resolve(self):
   resolve_accelerator_resources()
   resolve_memory_resources()
   resolve_labels()
   return self
```

cc @MengjinYan

ryanaoleary replied:
Done in [ed1636f](https://github.com/ray-project/ray/pull/53360/commits/ed1636f03fc7b00fd244bd900625b544da370de8), I moved the resolution logic outlined above to their own helpers and changed them to directly update the fields on `self` rather than returning a new `ResourceSpec`.

ryanaoleary committed:
Add ResourceAndLabelSpec unit test file and add helpers

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
remove change to test_basic_5

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Update python/ray/_private/resource_and_label_spec.py

Co-authored-by: Jiajun Yao <jeromeyjj@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary committed:
Update python/ray/_private/node.py

Co-authored-by: Jiajun Yao <jeromeyjj@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary committed:
Update python/ray/_private/resource_and_label_spec.py

Co-authored-by: Jiajun Yao <jeromeyjj@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary committed:
Update python/ray/_private/resource_and_label_spec.py

Co-authored-by: Jiajun Yao <jeromeyjj@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary committed:
Update python/ray/_private/resource_and_label_spec.py

Co-authored-by: Jiajun Yao <jeromeyjj@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary committed:
Update num-gpus being set and fix nits

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary commented on python/ray/_private/resource_and_label_spec.py:
```diff
@@ -0,0 +1,396 @@
+                f"contains {visible_accelerator_ids}."
+            )
+
+        if num_accelerators:
```
Oh yeah that's correct, we check that`num_accelerators > 0` before returning in the other function. Removed this check in 82c8558.

ryanaoleary commented on python/ray/_private/resource_and_label_spec.py:
```diff
@@ -0,0 +1,396 @@
+        self.object_store_memory = object_store_memory
+        self.memory = memory
+
+    def _get_current_node_accelerator(self) -> Optional[Tuple[AcceleratorManager, int]]:
```
Done in 82c8558.

ryanaoleary commented on python/ray/_private/resource_and_label_spec.py:
```diff
@@ -0,0 +1,396 @@
+        self.object_store_memory = object_store_memory
+        self.memory = memory
+
+    def _get_current_node_accelerator(self) -> Optional[Tuple[AcceleratorManager, int]]:
```
Done in 82c8558

ryanaoleary commented on python/ray/_private/resource_and_label_spec.py:
```diff
@@ -0,0 +1,396 @@
+            else:
+                self.resources[accelerator_resource_name] = num_accelerators
+                if self.num_gpus is None:
+                    self.num_gpus = 0
```
Oh I didn't realize we didn't want to add a "GPU" resource to `resources` if `num_gpus` is set but it's not found in `resources`. Fixed in 82c8558 to match how it was previously implemented [here](https://github.com/ray-project/ray/blob/c006d81a4b5681f50b684f6878b616c24574f000/python/ray/_private/resource_spec.py#L196). I also changed it to initialize `num_gpus` to 0 if it's `None` before this function is called which matches how it was implemented [here](https://github.com/ray-project/ray/blob/c006d81a4b5681f50b684f6878b616c24574f000/python/ray/_private/resource_spec.py#L154).

ryanaoleary commented on python/ray/_private/resource_and_label_spec.py:
```diff
@@ -0,0 +1,396 @@
+        """Detect and update accelerator resources on a node."""
+        if not accelerator_manager:
+            if self.num_gpus is None:
+                self.num_gpus = num_accelerators or 0
```
Yeah that's correct, fixed in [82c8558](https://github.com/ray-project/ray/pull/53360/commits/82c85586714f6751de234b9d33b731306227426a), we now set `num_gpus` to 0 if it's `None` in `resolve` before calling these functions.

ryanaoleary commented on python/ray/_private/resource_and_label_spec.py:
```diff
@@ -0,0 +1,396 @@
+        if accel_info:
+            accelerator_manager, num_accelerators = accel_info
+        else:
+            accelerator_manager, num_accelerators = None, 0
```
Done in [82c8558](https://github.com/ray-project/ray/pull/53360/commits/82c85586714f6751de234b9d33b731306227426a)

jjyao commented on python/ray/_private/resource_and_label_spec.py:
```diff
@@ -0,0 +1,405 @@
+
+        Returns:
+            Tuple[Optional[AcceleratorManager], int]: A tuple containing the accelerator
+            manager (or None) and the detected accelerator count (0 if none found).
```
We don't just return the deteced accelerator count in this function but a merge of what user specified and what's auto-detected. Could you update the comment to reflect that?

ryanaoleary replied:
Updated to be more descriptive in 9f3b03d.

jjyao commented on python/ray/_private/resource_and_label_spec.py:
```diff
@@ -0,0 +1,405 @@
+
+        Returns:
+            Tuple[Optional[AcceleratorManager], int]: A tuple containing the accelerator
+            manager (or None) and the detected accelerator count (0 if none found).
```
This function doesn't only return the number of detected accelerator but it's a merge of user configured number and auto-detected number. Could you update the comments to reflect that.

jjyao reviewed (commented):
LG! Could you also make sure for all the bugs we found in the review comments, there are tests for them?

ryanaoleary committed:
Remove test_sql.py change

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary commented on python/ray/data/tests/test_sql.py:
```diff
@@ -306,7 +306,10 @@ def request_get_mock(url, params=None, **kwargs):
             assert False, "Invalid request."
 
-        with mock.patch("requests.get", request_get_mock), mock.patch(
+        with mock.patch(
+            "ray._private.accelerators.tpu.TPUAcceleratorManager.get_current_node_accelerator_type",
+            return_value=None,
+        ), mock.patch("requests.get", request_get_mock), mock.patch(
```
The change to calling `get_current_node_accelerator_type` when we were detecting the accelerator type for default labels was causing one of the tests to fail with unexpected output in the stack trace. I don't think we need this change with the refactor so I'll remove it.

ryanaoleary replied:
Done in [9f3b03d](https://github.com/ray-project/ray/pull/53360/commits/9f3b03d2b7062c193c37183c18959a14a914766c) and verified it still passes.

ryanaoleary committed:
Fix GPU tests, use monkeypatch, and fix comment

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Fix num_gpus edge case and add additional unit tests

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary commented:
> LG! Could you also make sure for all the bugs we found in the review comments, there are tests for them?

Done with [b68babe](https://github.com/ray-project/ray/pull/53360/commits/b68babeb520c9bbc14cc24578a5447c5f74de604). I believe there are now unit tests to cover all the edge cases. I've verified the tests are passing and the resolution behavior is on par with the previous `ResourceSpec`.

ryanaoleary committed:
Move unit tests file, change to call public APIs

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary commented on python/ray/tests/test_resource_and_label_spec.py:
```diff
@@ -0,0 +1,201 @@
+from ray._private.resource_and_label_spec import ResourceAndLabelSpec
+from ray._common.constants import HEAD_NODE_RESOURCE_NAME, NODE_ID_PREFIX
+import ray._private.ray_constants as ray_constants
+
```
Done in [6e6db15](https://github.com/ray-project/ray/pull/53360/commits/6e6db1541f9d423e2eb561b077beab9962eca9d2).

ryanaoleary commented on python/ray/tests/test_resource_and_label_spec.py:
```diff
@@ -0,0 +1,201 @@
+    mock_get_mgr.return_value = mock_mgr
+
+    result = ResourceAndLabelSpec._get_current_node_accelerator(None, {})
+    assert result == (mock_mgr, 2)
```
Sounds good, I refactored the tests in [6e6db15](https://github.com/ray-project/ray/pull/53360/commits/6e6db1541f9d423e2eb561b077beab9962eca9d2) to check the same edge cases but call public APIs like `resolve()`.

ryanaoleary commented:
> @ryanaoleary there are test failures, could you take a look?

Fixed in [6e6db15](https://github.com/ray-project/ray/pull/53360/commits/6e6db1541f9d423e2eb561b077beab9962eca9d2). `test_runtime_context` needed to be updated for the default labels and naming change of `ray.io/node_id` -> `ray.io/node-id`.

jjyao commented on python/ray/tests/unit/BUILD:
```diff
@@ -26,7 +26,10 @@ py_test_run_all_subdirectory(
     size = "small",
-    srcs = ["test_runtime_env_validation.py"],
+    srcs = [
+        "test_resource_and_label_spec.py",
```
No need to add it here, the previous bazel target covers it.

ryanaoleary replied:
[cfba0b5](https://github.com/ray-project/ray/pull/53360/commits/cfba0b5d8df68ac31f340297e9b6c96736670438)

jjyao commented on python/ray/tests/unit/test_resource_and_label_spec.py:
```diff
@@ -0,0 +1,147 @@
+    assert spec.num_gpus == 0
+    assert isinstance(spec.labels, dict)
+    assert HEAD_NODE_RESOURCE_NAME in spec.resources
+    assert any(key.startswith(NODE_ID_PREFIX) for key in spec.resources.keys())
```
check memory as well?

ryanaoleary replied:
Added two tests related to memory in https://github.com/ray-project/ray/commit/b6b2799fa697b9f52ac27721ca7be297807bf67e.

jjyao commented on python/ray/tests/unit/test_resource_and_label_spec.py:
```diff
@@ -0,0 +1,147 @@
+    monkeypatch.setenv(ray_constants.LABELS_ENVIRONMENT_VARIABLE, "{}")
+
+    with patch("ray._private.utils.get_num_cpus", return_value=4), patch(
+        "ray.util.get_node_ip_address", return_value="127.0.0.1"
+    ), patch(
+        "ray._private.utils.estimate_available_memory", return_value=1_000_000_000
+    ), patch(
+        "ray._common.utils.get_system_memory", return_value=2_000_000_000
+    ), patch(
+        "ray._private.utils.get_shared_memory_bytes", return_value=50_000
+    ):
```
If we don't check them, there is no need to patch them

ryanaoleary replied:
I think for some of these (the memory related ones that get called in `_resolve_memory_resources`) we do actually need them so that we can verify that value that gets resolved. I'll remove the ones that are unused/don't get checked.

jjyao commented on python/ray/tests/unit/test_resource_and_label_spec.py:
```diff
@@ -0,0 +1,147 @@
+    )
+
+    with patch("ray._private.utils.get_num_cpus", return_value=1), patch(
+        "ray.util.get_node_ip_address", return_value="127.0.0.1"
+    ), patch(
+        "ray._private.utils.estimate_available_memory", return_value=1_000_000
+    ), patch(
+        "ray._common.utils.get_system_memory", return_value=2_000_000
+    ), patch(
+        "ray._private.utils.get_shared_memory_bytes", return_value=50_000
+    ):
```
No need to patch, we don't test them.

ryanaoleary replied:
These functions get called from resolve() so I wanted to have them return a known value, so that when we later check the values set in each field we know what to expect. I removed any that are unused/we don't test in https://github.com/ray-project/ray/commit/b6b2799fa697b9f52ac27721ca7be297807bf67e.

jjyao commented on python/ray/tests/unit/test_resource_and_label_spec.py:
```diff
@@ -0,0 +1,147 @@
+        spec = ResourceAndLabelSpec()
+        spec.resolve(is_head=False)
+
+    assert spec.num_gpus == 4
```
Do we have another test where auto-detected is 4 but manually set one is 2 and we should use 2.

In general, could you make sure the tests can cover all code path.

jjyao reviewed (commented):
Looks great! Just some comments on the tests.

MengjinYan reviewed (commented):
Mostly nit comments. Almost there!

ryanaoleary committed:
Update python/ray/_private/resource_and_label_spec.py

Co-authored-by: Mengjin Yan <mengjinyan3@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary committed:
Update python/ray/_private/resource_and_label_spec.py

Co-authored-by: Mengjin Yan <mengjinyan3@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary commented on python/ray/_private/resource_and_label_spec.py:
```diff
@@ -2,75 +2,67 @@
+        self.resources = resources
+        self.labels = labels
+
+    def resolved(self) -> bool:
```
That makes sense, I'll add that flag and change the implementation `resolved()` to just check that the flag was set at the end of the `resolve()` call. I'll move the old implementation of `resolved()` to its own function, since we still want to call that at the end of `resolve()` to ensure that all required fields are not `None`.

ryanaoleary committed:
Add additional unit tests and move resource merge/load from env to ResourceAndLabelSpec

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary commented on python/ray/_private/resource_and_label_spec.py:
```diff
@@ -2,75 +2,67 @@
 
     All fields can be None. Before starting services, resolve() should be
     called to return a ResourceAndLabelSpec with unknown values filled in with
     defaults based on the local machine specifications.
```
https://github.com/ray-project/ray/commit/b6b2799fa697b9f52ac27721ca7be297807bf67e

ryanaoleary committed:
Make functions that don't use self static

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Remove change to unit/BUILD

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

ryanaoleary committed:
Format

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

jjyao commented on python/ray/tests/unit/test_resource_and_label_spec.py:
```diff
@@ -0,0 +1,361 @@
+        spec.resolve(is_head=False)
+
+
+@patch("ray._private.usage.usage_lib.record_hardware_usage", lambda *_: None)
```
I don't think you need this patch

ryanaoleary replied:
8bb604b

jjyao commented on python/ray/tests/unit/test_resource_and_label_spec.py:
```diff
@@ -0,0 +1,361 @@
+        "ray._private.accelerators.get_all_accelerator_resource_names",
+        return_value=["GPU"],
+    ).start()
+    mock_mgr.get_current_node_num_accelerators.return_value = 4
+    mock_mgr.get_current_process_visible_accelerator_ids.return_value = [
+        "0",
+        "1",
+        "2",
+        "3",
+    ]
```
Instead of mocking, can we define a FakeGPUAcceleratorManager that implements AcceleratorManager and return it from `get_accelerator_manager_for_resource`

jjyao commented on python/ray/tests/unit/test_resource_and_label_spec.py:
```diff
@@ -0,0 +1,361 @@
+    assert spec.resources.get("accelerator_type:A100") == 1
+
+
+@patch("ray._private.usage.usage_lib.record_hardware_usage", lambda *_: None)
```
No need for it.

jjyao commented on python/ray/tests/unit/test_resource_and_label_spec.py:
```diff
@@ -0,0 +1,361 @@
+        "ray._private.accelerators.get_all_accelerator_resource_names",
+        return_value=["GPU"],
+    ).start()
+    mock_mgr.get_current_node_num_accelerators.return_value = 4
+    mock_mgr.get_current_process_visible_accelerator_ids.return_value = [
+        "0",
+        "1",
+        "2",
+        "3",
+    ]
```
Change all mocks to use fake accelerator manager, same as above

jjyao commented on python/ray/tests/unit/test_resource_and_label_spec.py:
```diff
@@ -0,0 +1,361 @@
+    assert spec.resources.get("accelerator_type:A100") == 1
+
+
+@patch("ray._private.usage.usage_lib.record_hardware_usage", lambda *_: None)
```
del

jjyao reviewed (approved):
Nice tests!

ryanaoleary committed:
Update python/ray/_private/resource_and_label_spec.py

Co-authored-by: Jiajun Yao <jeromeyjj@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary committed:
Update python/ray/_private/resource_and_label_spec.py

Co-authored-by: Jiajun Yao <jeromeyjj@gmail.com>
Signed-off-by: Ryan O'Leary <113500783+ryanaoleary@users.noreply.github.com>

ryanaoleary commented on python/ray/tests/unit/test_resource_and_label_spec.py:
```diff
@@ -0,0 +1,361 @@
+    assert any(key == "autoscaler-override-label" for key in spec.labels)
+
+
+@patch("ray._private.usage.usage_lib.record_hardware_usage", lambda *_: None)
```
Without the patch the tests were failing with:
```
hardware_usage = 'TPU-v6e'

    def record_hardware_usage(hardware_usage: str):
        """Record hardware usage (e.g. which CPU model is used)"""
>       assert _internal_kv_initialized()
E       AssertionError

../../_private/usage/usage_lib.py:269: AssertionError
```
since we call:
```
usage_lib.record_hardware_usage(accelerator_type)
```
when setting the `accelerator-type` in `_resolve_accelerator_resources`.

ryanaoleary replied:
Done in 8bb604b.

MengjinYan commented on python/ray/_private/resource_and_label_spec.py:
```diff
@@ -119,19 +125,36 @@ def to_resource_dict(self):
-            self.resources = {}
+        # Load environment override resources and merge with resources passed
+        # in from Ray Params. Separates special case params if found in env.
+        env_resources = ResourceAndLabelSpec._load_env_resources()
```
Nit: It seems that the user input resources are resolved here while the user input labels are resolved in the `_resolve_labels` function. Wondering should we put both of them here or we can put both of them in their functions corresponding functions?

MengjinYan reviewed (approved):
Just 1 super nit comment! Thanks for following through!

ryanaoleary committed:
Change tests to use FakeAcceleratorManager, add _resolve_resources, and move record hardware usage to node.py

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>

jjyao reviewed (approved):
Nice!

MengjinYan reviewed (approved):
Thanks!

MengjinYan commented:
@ryanaoleary Looks like there is a conflict needs to be resolved.

ryanaoleary committed:
Fix import of usage_lib

Signed-off-by: Ryan O'Leary <ryanaoleary@google.com>
